gh-69714: Make `calendar` module fully tested

bxsx opened the pull request:
# Summary

Resolves #69714 by increasing test coverage of `calendar` module to _100%_[1].

## Coverage reports

### Before
`main` branch test coverage report for `calendar` module:
```
$ ./python.exe coveragepy report --show-missing
Name              Stmts   Miss  Cover   Missing
-----------------------------------------------
Lib/calendar.py     399     60    85%   559, 564-570, 582, 602, 664-764, 768
-----------------------------------------------
TOTAL               399     60    85%
```

### After
The PR branch test coverage report for `calendar` module:
```
$ ./python.exe coveragepy report --show-missing
Name              Stmts   Miss  Cover   Missing
-----------------------------------------------
Lib/calendar.py     397      3    99%   722, 742, 766
-----------------------------------------------
TOTAL               397      3    99%
```

## Footnotes

[1] The remaining three lines are:
```
$ cat -n Lib/calendar.py | sed -n '722p;742p;766p'
   722	        sys.exit(1)
   742	            sys.exit(1)
   766	    main()
$
```
and they are, _in fact_, already covered by tests (only not reported by coveragepy).

<!-- gh-issue-number: gh-69714 -->
* Issue: gh-69714
<!-- /gh-issue-number -->

jesstess committed:
gh-57539: Add tests for `LocaleTextCalendar.formatweekday`

This patch has already been applied to `main` branch (via gh-93468), but
with wrong copyright. After merging this commit to `main`, git `Author`
metadata will be updated to the original author.

rohitmediratta committed:
Increased test coverage of calendar module

bxsx committed:
gh-69714: Restore the test thrown out by cb0523e67c

bxsx committed:
gh-69714: Increase test coverage for `calendar.Locale{Text|HTML}Calendar(locale='')`

bxsx committed:
gh-69714: Add missing test cases for custom locale in `calendar.Locale{Text|HTML}Calendar`

There are 3 paths to use `locale` argument in
`calendar.Locale{Text|HTML}Calendar.__init__(..., locale=None)`:
(1) `locale=None` -- denotes the "default locale"[1]
(2) `locale=""` -- denotes the native environment
(3) `locale=other_valid_locale` -- denotes a custom locale

So far case (2) is covered and case (1) is in 78935daf5a (same branch).
This commit adds a remaining case (3).

[1] In the current implementation, this translates into the following
approach:

GET current locale
IF current locale == "C" THEN
  SET current locale TO ""
  GET current locale
ENDIF

bxsx committed:
gh-69714: Remove unreachable code (and increase test coverage)

This condition cannot be true. `_locale.setlocale()` from the C module
raises `locale.Error` instead of returning `None` for
`different_locale.__enter__` (where `self.oldlocale` is set).

bxsx committed:
gh-69714: Increase CLI tests coverage

bxsx committed:
gh-69714: Add more test cases to `calendar` cmdline

bxsx committed:
gh-69714: Increase test coverage for `calendar` cmdline

bxsx committed:
gh-69714: Increase test coverage for `LocaleTextCalendar.formatmonthname`

bxsx committed:
Merge branch 'main' into gh-69714/calendar-test-coverage

AA-Turner commented on Lib/calendar.py:
```diff
@@ -555,8 +555,6 @@ def __enter__(self):
         _locale.setlocale(_locale.LC_TIME, self.locale)
 
     def __exit__(self, *args):
-        if self.oldlocale is None:
-            return
```
Can you explain this change?

A

bxsx replied:
I explained it in the commit message. See a96786c.

AA-Turner replied:
This was introduced in @vstinner's #31214 -- I agree with you though, `locale.setlocale` calls through to `_locale.setlocale` which appears to only return strings.

A

bxsx replied:
I looked that PR up and believe these lines were added by mistake, following the common pattern of using the context manager. However, @vstinner mentions "corner cases" in https://github.com/python/cpython/pull/31214#issuecomment-1049863219, so maybe there is more to it. @vstinner can you comment?

In any case, if you think a96786c should be reverted, I'm fine with it. This is just a slight increase in the test coverage. BTW. if there is any corner case known, maybe we could cover it as well?

vstinner replied:
I wrote the test to handle the case when ``self.oldlocale = _locale.setlocale(_locale.LC_TIME, None)`` fails in ``__enter__()``. I tested. It seems like ``__exit__()`` is not called if ``__enter__()`` raises an exception. So removing "oldlocale is None" test is correct.

bxsx replied:
Thanks for the clarification! I'm marking this as resolved.

> It seems like `__exit__()` is not called if `__enter__()` raises an exception.

I believe this is part of the context manager protocol (`__exit__()` is only called if `__enter__()` has successfully returned the resource), however, I cannot find it in PEP-343 right now. But yeah, I'm pretty sure this is how it works since the `with` statement introduction in Python 2.5.

AA-Turner commented on Lib/calendar.py:
```diff
@@ -660,7 +658,7 @@ def timegm(tuple):
 
 
-def main(args):
+def main(args=None):
```
Can you explain the changes to `main()` -- whilst I believe they should have the same behaviour, changing this should ideally be done in another PR.

A

bxsx replied:
> Can you explain the changes to `main()`

The reason to make the change is explained in the commit message: 636b4e1

> whilst I believe they should have the same behaviour, changing this should ideally be done in another PR.

Well, the major factor in increasing test coverage by 14% is only possible with this change. I believe that this is a core part of this PR. I can split this PR into several smaller PRs, but this change is related to about 80% of the lines added anyway..

AA-Turner replied:
The alternative approach would be to hijack `sys.argv` either via monkeypatching or restoring the value at the end, but that is messy. Can you keep `main(sys.argv)`?

A

bxsx replied:
> Can you keep `main(sys.argv)`?

I'm afraid I can't without an additional conditional code. However, I think `main(sys.argv[1:])` might work. Can you explain why it's better to keep the old way?

I think my change is compliant with the `argparse` protocol and also backwards compatible (it uses `sys.argv` by default).

See:
https://github.com/python/cpython/blob/30610d28374f5a9698d456cebf3ae496ac01af51/Lib/argparse.py#L1861-L1874

AA-Turner replied:
> Can you explain why it's better to keep the old way?

Easier to review a PR that does less, especially if it touches less 'real' stdlib code. But as you point out it seems it'd have to change anyway so might as well just call `main()`.

A

AA-Turner commented on Lib/test/test_calendar.py:
```diff
@@ -546,21 +548,60 @@ def test_locale_calendars(self):
+        local_weekday = cal.formatweekday(1)
+        local_month = cal.formatmonthname(2010, 10)
+        self.assertIsInstance(local_weekday, str)
+        self.assertIsInstance(local_month, str)
```
These could be new test methods -- the original `test_locale_calendars` noted it was just to "ensure that Locale{Text,HTML}Calendar resets the locale properly". The additions here go beyond ensuring that the locale isn't mutated post-hoc.

bxsx replied:
Yes, that was my first impression too. But somehow, I couldn't find any comment on this in the original review (see [BPO-25528](https://bugs.python.org/issue25528)) of the patch.
I also noticed that the name of the test case is generic and allows you to add more cases in the same place. So it looks like the comment rather justifies why `old_october` is ever assigned.

I will add a dedicated test case to _"ensure that Locale{Text,HTML}Calendar resets the locale properly"_.

AA-Turner commented on Lib/test/test_calendar.py:
```diff
@@ -577,6 +618,20 @@ def test_locale_calendar_formatweekday(self):
+            # formatmonthname uses the same month names regardless of the width argument.
+            cal = calendar.LocaleTextCalendar(locale='en_US')
+        except locale.Error:
+            # cannot set the system default locale -- skip rest of test
```
```suggestion
```

AA-Turner reviewed (commented):
Also cc: @hugovk (who I know has different opinions on testing to me!)

A

bxsx committed:
Remove comment

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

bxsx committed:
gh-69714: Reorder locale in the test case

bxsx committed:
gh-69714: Extract test case

bxsx committed:
Update ACKS

bxsx committed:
Add NEWS entry

bxsx committed:
Revert "gh-69714: Remove unreachable code (and increase test coverage)"

This reverts commit a96786cc1fe9de9f051884ed66e75b87c7fde1ba.

bxsx committed:
Expand the try clause to calls to `LocaleTextCalendar.formatmonthname()`.

This method temporarily changes the current locale to the given locale,
so `_locale.setlocale()` may raise `local.Error`.

bxsx committed:
Revert "Revert "gh-69714: Remove unreachable code (and increase test coverage)""

This reverts commit 40e0da34fdc0a7a3484039fd37ae0e0d68d157c1.
So it brings back commit a96786cc1fe9de9f051884ed66e75b87c7fde1ba.

bxsx committed:
Add missing whitespace

hugovk commented on Lib/test/test_calendar.py:
```diff
@@ -577,6 +640,19 @@ def test_locale_calendar_formatweekday(self):
+            # For too short widths, a full name (with year) is used.
+            self.assertEqual(cal.formatmonthname(2022, 6, 2, withyear=False), "June")
+            self.assertEqual(cal.formatmonthname(2022, 6, 2, withyear=True), "June 2022")
+            # For long widths, a centered name is used.
```
In addition to even widths, maybe include some odd widths?

bxsx replied:
Thanks! Added in [0aa0caf](https://github.com/python/cpython/pull/93655/commits/0aa0caf8b8e67de0f3e222c697f4cd06c3061974)

hugovk commented on Lib/test/test_calendar.py:
```diff
@@ -835,46 +911,110 @@ def conv(s):
         return assert_python_ok('-m', 'calendar', *args)[1]
 
-    def assertFailure(self, *args):
+    def assertCLIFails(self, *args):
```
```suggestion
    def assertCliFails(self, *args):
```

bxsx replied:
Let's stay in line with:
> _When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError._ (from [PEP-8](https://peps.python.org/pep-0008/#descriptive-naming-styles))

hugovk commented on Lib/test/test_calendar.py:
```diff
@@ -542,25 +544,86 @@ def test_months(self):
+        self.assertIsInstance(local_month, str)
+        self.assertEqual(len(local_weekday), 10)
+        self.assertEqual(len(local_weekday_abbr), 3)
+        self.assertGreaterEqual(len(local_month), 10)
```
Personally, rather than one long test method that covers many cases (multiple blocks of setups and asserts), I prefer to keep each method short: test one thing.

That way, if a middle case fails, the remainder of the cases still run.

(If necessary, create a helper to make the skip bit easier.)

bxsx replied:
Agreed. However, this particular test case takes advantage of switching between locales/another case.
So the design is correct. However, as the code gets a bit larger, we can split the cases of `LocaleTextCalendar` and `LocaleHTMLCalendar` (e73fc9e).

bxsx committed:
gh-69714: Move the validation to the beginning of the function

No need to process the function.
This also helps in supporting CLI testing.

bxsx committed:
gh-69714: Increase test coverage for illegal arguments

bxsx committed:
Revert "gh-69714: Add more test cases to `calendar` cmdline"

These tests are now part of `test_illegal_arguments`.
This reverts commit 238c5271226f500b4b395dfcd8d05aa998870390.

bxsx committed:
Clean up redundant arguments

bxsx committed:
gh-69417: Include cases with odd widths in `formatmonthname`

bxsx committed:
gh-69714: Split tests for LocaleTextCalendar and LocaleHTMLCalendar

AA-Turner commented on Lib/calendar.py:
```diff
@@ -726,6 +724,9 @@ def main(args):
 
     if options.type == "html":
+        if options.month:
+            parser.error("incorrect number of arguments")
```
This could include a suggestion to the user on what to change?
```suggestion
            parser.error("Incorrect number of arguments.")
```

bxsx replied:
While this sounds reasonable, it's not the actual point of the error message.
You can imagine it as an `argparse` extension, therefore the error message should be in the same format (your suggested change breaks it) and in line with other `argparse` errors (to answer your question).

Also, I think this is a bit outside the scope of this PR, and in case we would like to do such a change, should it probably be done in a separate PR?

AA-Turner replied:
You're right on two counts -- I didn't look for enough context on this one.

A

AA-Turner commented on Lib/calendar.py:
```diff
@@ -726,6 +724,9 @@ def main(args):
     if options.type == "html":
+        if options.month:
+            parser.error("incorrect number of arguments")
+            sys.exit(1)
```
And on line 722
```suggestion
            raise SystemExit(1)
```

bxsx replied:
This is outside of the scope of this PR.
BTW. May I ask why you want this change?

AA-Turner replied:
Second count of my error, I had thought you'd added the `sys.exit` call. As to why, it saves indirection and a function call -- `sys.exit` uses `SystemExit` internally. You can see that it takes fewer instructions in `dis` output:

```pycon
>>> def f(): sys.exit(1)
>>> def g(): raise SystemExit(1)
>>> dis.dis(f)
  1           0 LOAD_GLOBAL              0 (sys)
              2 LOAD_METHOD              1 (exit)
              4 LOAD_CONST               1 (1)
              6 CALL_METHOD              1
              8 POP_TOP
             10 LOAD_CONST               0 (None)
             12 RETURN_VALUE
>>> dis.dis(g)
  1           0 LOAD_GLOBAL              0 (SystemExit)
              2 LOAD_CONST               1 (1)
              4 CALL_FUNCTION            1
              6 RAISE_VARARGS            1
```

A

bxsx replied:
Thanks. While that makes sense, isn't it rather premature optimization?

AA-Turner replied:
Knuth would be disappointed!

A

AA-Turner reviewed (commented):
The updated tests look good, I'll wait for Hugo's thoughts.

A

iritkatriel committed:
Merge branch 'main' into gh-69714/calendar-test-coverage

bxsx committed:
Merge branch 'main' into gh-69714/calendar-test-coverage

bxsx committed:
Merge branch 'main' into gh-69714/calendar-test-coverage

StanFromIreland commented on Lib/calendar.py:
```diff
@@ -756,6 +754,9 @@ def main(args):
     locale = options.locale, options.encoding
 
     if options.type == "html":
+        if options.month:
```
Edit: misread
